### PR TITLE
Add "client-metrics" service

### PR DESCRIPTION
**Pull request by Marketen**
Client Metrics is a new service that exposes staker package metrics through an Express API on port 9090. The service has a single endpoint, "/metrics," which is used by the Prometheus server on DMS to fetch all its metrics.

Staker package metrics are derived from the JSON-RPC APIs provided by the Ethereum packages. The specifications for these APIs can be found at:
- https://ethereum.github.io/execution-apis/api-documentation/ for execution clients
- https://ethereum.github.io/beacon-APIs/#/ for consensus clients

Currently, the service only includes two metrics, `api_rpc_exec_syncing` and `api_rpc_cons_syncing`. These metrics are build from the JSON-RPCI API calls that staking clients offer.

Since each dappnode can have up to three different active execution and consensus clients in the Stakers UI, the metrics are fetched from each of them. If a staker package is down, the metric from that package is not returned.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,5 @@
 # DAppNodeSDK release directories
 build
-build_*
\ No newline at end of file
+build_*
+node_modules
```
Comment (pablomendezroyo) [inline .gitignore:4]: Does not node_modules belongs to `client-metrics` dir?

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,5 @@
...
+dist
```
Comment (pablomendezroyo) [inline .gitignore:5]: Does not dist belongs to `client-metrics` dir?

File: client-metrics/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,21 @@
+FROM node:18-alpine
+
+# Set the working directory inside the container
+WORKDIR /app
+
+# Copy package.json, tsconfig.json and yarn.lock (if present) to the container
+COPY package.json yarn.lock* tsconfig.json ./
+
+# Install dependencies using yarn
+RUN yarn install
```
Comment (pablomendezroyo) [inline client-metrics/Dockerfile]: I encourage you to only install production libraries to reduce the final application size

File: client-metrics/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+# Copy the source code to the container
+COPY ./src ./src
+
+# Build the TypeScript code
+RUN yarn build
+
+EXPOSE 3002
+
+# Start the Express server
+CMD ["node", "./dist/index.js"]
```
Comment (pablomendezroyo) [inline client-metrics/Dockerfile]: The above docker instructions should belong to a build stage, and you should add a final stage for production by copying only the needed files

File: client-metrics/src/stakerConfig.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    case "besu.public.dappnode.eth":
+      executionClientMainnetUrl = `http://besu.public.dappnode:8545`;
+      break;
+    case "nethermind.public.dappnode.eth":
+      executionClientMainnetUrl = `http://nethermind.public.dappnode:8545`;
+      break;
+    case "erigon.dnp.dappnode.eth":
+      executionClientMainnetUrl = `http://erigon.dappnode:8545`;
+      break;
+   //TODO: add error handling, default case
```
Comment (pablomendezroyo) [inline client-metrics/src/stakerConfig.ts]: The var may take empty strin value and its not an error, just no client configured for the given network

File: client-metrics/src/stakerConfig.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  ExecutionClientMainnet,
+  ConsensusClientMainnet,
+  ExecutionClientGnosis,
+  ExecutionClientPrater,
+  ConsensusClientPrater,
+  ConsensusClientGnosis,
+
+} from "@dappnode/types";
+
+export function convertEnvsToURL() {
```
Comment (pablomendezroyo) [inline client-metrics/src/stakerConfig.ts]: I encourage you to add to the `tsconfig.json` the lint rule `explicitReturn` this way you are forced to add the return type of the function and if it does not match you will get a lint error.

File: client-metrics/src/stakerConfig.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        break;
+      case "teku-gnosis.dnp.dappnode.eth":
+        consensusClientGnosisUrl = `http://beacon-chain.teku-gnosis.dappnode:3500`;
+        break;
+      case "lodestar-gnosis.dnp.dappnode.eth":
+        consensusClientGnosisUrl = `http://beacon-chain.lodestar.dappnode:3500`;
+        break;
+    }
+
+  return {
```
Comment (pablomendezroyo) [inline client-metrics/src/stakerConfig.ts]: The return type will be a string or undefined so for cases where the value is empty string you return undefined

File: client-metrics/src/metrics.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    id: 0,
+  });
+
+  const executionClientUrls = {
+    mainnet: urls.executionClientMainnetUrl,
+    prater: urls.executionClientPraterUrl,
+    gnosis: urls.executionClientGnosisUrl,
+  };
+
+  const networks = ["mainnet", "prater", "gnosis"];
```
Comment (pablomendezroyo) [inline client-metrics/src/metrics.ts]: you should be able to get the networks available using the @dappnode/types library.

See https://github.com/dappnode/types/blob/ac36a72059103d5e2e16dc0857d21f86abcdea57/src/types/stakers.ts#L2

File: client-metrics/src/metrics.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+// Function to make the axios call
+async function jsonRPCapiCall(APImethod: string, params?: string[]) {
+  const data = JSON.stringify({
+    jsonrpc: "2.0",
+    method: APImethod,
+    params: params,
+    id: 0,
+  });
+
+  const executionClientUrls = {
```
Comment (pablomendezroyo) [inline client-metrics/src/metrics.ts]: Execution client url may or may not exist dependeing if the env is used or not (see below commands). I would suggest to filter those !== undefined or just return the ones that exist in the function `convertEnvsToURL`

File: client-metrics/src/metrics.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
+import { convertEnvsToURL } from "./stakerConfig.js";
+import promClient from "prom-client";
+import axios from "axios";
+import express from "express";
+
+const register = new promClient.Registry();
+const urls = convertEnvsToURL()
+
+// Function to make the axios call
+async function jsonRPCapiCall(APImethod: string, params?: string[]) {
```
Comment (pablomendezroyo) [inline client-metrics/src/metrics.ts]: See comment below, you should add to the tsconfig the return function lint rule

File: client-metrics/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,6 @@
+import { app } from "./metrics.js";
+
+const PORT = 3002;
```
Comment (pablomendezroyo) [inline client-metrics/src/index.ts]: Its considered a standard to use the port 9090 for metrics

File: client-metrics/src/metrics.ts (hunk 10 above)
Comment (pablomendezroyo) [inline client-metrics/src/metrics.ts]: For me makes more sense to add the arg `network` to this function and then when building the metric you can simply do something like:
```typescript
await Promise.all(["mainnet","prater","gnosis"].map(network => {
  // here you know the network to set the metric
  const result = await jsonRPCapiCall(...)...
}))
```

File: client-metrics/src/metrics.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        headers: {
+          "Content-Type": "application/json",
+        },
+        data: data,
+      };
+
+      try {
+        const response = await axios(config);
+        return { network, response: response.data.result };
+      } catch (error) {
```
Comment (pablomendezroyo) [inline client-metrics/src/metrics.ts]: Do not omit the error at least print it for future debugging

File: client-metrics/src/metrics.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        url: executionClientUrl,
+        headers: {
+          "Content-Type": "application/json",
+        },
+        data: data,
+      };
+
+      try {
+        const response = await axios(config);
+        return { network, response: response.data.result };
```
Comment (pablomendezroyo) [inline client-metrics/src/metrics.ts]: its weird to return the network at this point, see the comment from above. This function should be agnostic of the network as long as the execution clients have the same api standard

File: client-metrics/src/metrics.ts (hunk 12 above)
Comment (pablomendezroyo) [reply client-metrics/src/metrics.ts]: Also, consider dont doing any try-catch here but doing the error handling when setting the metric, so this function remains as pure as possible.

File: client-metrics/src/stakerConfig.ts [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+  const ConsensusClientGnosis = process.env._DAPPNODE_GLOBAL_CONSENSUS_CLIENT_GNOSIS as ConsensusClientGnosis;
+
+  let executionClientMainnetUrl: string,
+    executionClientPraterUrl: string,
+    executionClientGnosisUrl: string,
+    consensusClientMainnetUrl: string,
+    consensusClientPraterUrl: string,
+    consensusClientGnosisUrl: string;
+
+  switch(executionClientMainnet) {
```
Comment (pablomendezroyo) [inline client-metrics/src/stakerConfig.ts]: This code should not be placed here but in the npm dappnode package @dappnode/types.

Could you open an issue in that repository (something like "create utility function `getExecutionEndpointFromDnpName`" and same for consensus) and leave a coment here refereing the issue please

File: client-metrics/tsconfig.json [hunk 15]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (pablomendezroyo) [inline client-metrics/tsconfig.json]: you should add more strict rules

File: .gitignore [hunk 16]
```diff
@@ -1,3 +1,5 @@
 # DAppNodeSDK release directories
```
Comment (pablomendezroyo) [inline .gitignore:1]: Could you please add to the project eslint configuration? `.eslintr.json`. Take @dappnode/types as reference

**Review by pablomendezroyo — CHANGES_REQUESTED** [review]

File: client-metrics/src/stakerConfig.ts (hunk 14 above)
Comment (Marketen) [reply client-metrics/src/stakerConfig.ts]: https://github.com/dappnode/types/issues/17

File: .gitignore (hunk 1 above)
Comment (Marketen) [reply .gitignore:4]: Yes! But node_modules ignores all "node_modules" directories in the current dir and subdirectories. 
https://stackoverflow.com/questions/29820791/git-ignore-node-modules-folder-everywhere

(It is ignoring my node modules in /src/ dir)

File: .gitignore (hunk 2 above)
Comment (Marketen) [reply .gitignore:5]: same as "node_modules", it works on current and sub directories

**Comment by pablomendezroyo** [comment]
@Marketen resolve conflicts

File: client-metrics/package.json [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+    "@dappnode/types": "^0.1.23",
+    "@types/node": "^20.2.4",
+    "@types/express": "^4.17.17",
+    "ts-node": "^10.9.1",
+    "axios": "^1.4.0",
+    "express": "^4.18.2",
+    "prom-client": "^14.1.0",
+    "eslint": "^8.39.0",
+    "eslint-plugin-import": "^2.27.5",
+    "@typescript-eslint/eslint-plugin": "^5.59.2",
```
Comment (pablomendezroyo) [inline client-metrics/package.json]: this should be a dev dependency

File: client-metrics/package.json [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+    "@typescript-eslint/parser": "5.0.0"
```
Comment (pablomendezroyo) [inline client-metrics/package.json]: this should be a dev dependency

File: client-metrics/package.json [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+  "dependencies": {
+    "@dappnode/types": "^0.1.23",
+    "@types/node": "^20.2.4",
+    "@types/express": "^4.17.17",
+    "ts-node": "^10.9.1",
+    "axios": "^1.4.0",
+    "express": "^4.18.2",
+    "prom-client": "^14.1.0",
+    "eslint": "^8.39.0",
+    "eslint-plugin-import": "^2.27.5",
```
Comment (pablomendezroyo) [inline client-metrics/package.json]: this should be a dev dependency

File: client-metrics/package.json [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "dependencies": {
+    "@dappnode/types": "^0.1.23",
+    "@types/node": "^20.2.4",
+    "@types/express": "^4.17.17",
+    "ts-node": "^10.9.1",
+    "axios": "^1.4.0",
+    "express": "^4.18.2",
+    "prom-client": "^14.1.0",
+    "eslint": "^8.39.0",
```
Comment (pablomendezroyo) [inline client-metrics/package.json]: this should be a dev dependency

File: client-metrics/package.json [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+    "@types/express": "^4.17.17",
+    "@types/node": "^20.2.4",
+    "ts-node": "^10.9.1",
+    "typescript": "^5.0.4"
+  },
+  "dependencies": {
+    "@dappnode/types": "^0.1.23",
+    "@types/node": "^20.2.4",
+    "@types/express": "^4.17.17",
+    "ts-node": "^10.9.1",
```
Comment (pablomendezroyo) [inline client-metrics/package.json]: this should be a dev dependency

File: client-metrics/src/metrics.ts [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+  const metric = type === "execution" ? executionSyncingMetric : consensusSyncingMetric;
+  metric.set({ network: network }, isSyncing);
+}
+
+register.registerMetric(consensusPeerCountMetric);
+register.registerMetric(executionPeerCountMetric);
+register.registerMetric(executionSyncingMetric);
+register.registerMetric(consensusSyncingMetric);
+
+const app = express();
```
Comment (pablomendezroyo) [inline client-metrics/src/metrics.ts]: I would move this and below code to the `index.ts` for organization

**Review by pablomendezroyo — CHANGES_REQUESTED** [review]

**Review by pablomendezroyo — APPROVED** [review]

**Pull request merged**